enhancements/authentication/unauthenticated-subjects.md

```diff
@@ -0,0 +1,137 @@
+  * Both kube-apiserver and openshift-apiserver create a set of cluster roles and cluster role bindings during bootstrap.
+    * This [link](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings) has the list of default cluster roles and cluster role bindings in Kubernetes.
+    * This [link](https://docs.openshift.com/container-platform/4.4/authentication/using-rbac.html#default-roles_using-rbac) has the list of default cluster roles in OpenShift.
+  * `rbac.authorization.kubernetes.io/autoupdate=true` annotation is also added to default policy resources.
+  * On start-up both openshift-apiserver and kube-apiserver reconcile their default policy resources as follows
+    * if the annotation is `rbac.authorization.kubernetes.io/autoupdate=true`
+	  * add missing permissions to cluster roles.
+	  * add missing subjects to cluster role bindings.
+    * if the annotation is deleted
+	  * add annotation to the RBAC resource.
+	  * add missing permissions to cluster roles.
+	  * add missing subjects to cluster role bindings.
+    * if the annotation is `rbac.authorization.kubernetes.io/autoupdate=false`
+	  * skips reconciliation of resource.
+  * reconciliation does not delete any existing permissions or subjects.
+  * An admin to opt-out of reconciliation for a specific resource by setting annotation `rbac.authorization.kubernetes.io/autoupdate=false`.
+  * Detailed behavior can be found in this [link](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#auto-reconciliation).
+
+* Openshift-apiserver operator implements a new controller to the reconcile following cluster role bindings created by openshift-apiserver.
+  * cluster-status-binding
+  * discovery
+  * system:openshift:discovery
+  * basic-users
+* If the annotation `rbac.authorization.kubernetes.io/autoupdate` is present and set to `true`, the controller removes the unauthenticated subject from the list of subjects and updates the resource in the API server.
+* Customers can opt-out of this reconciliation for a cluster role binding by setting the annotation `rbac.authorization.kubernetes.io/autoupdate` to `false`.
+
+### User Stories [optional]
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+### Risks and Mitigations
+
+* Because of OpenShift 4.1 allowing access to discovery endpoint, some customer implementations might be expecting this privileged access to anonymous users. Removing unauthenticated subject on an upgrade might break some of these implementations.
+* Informing customers beforehand of this change would help the customer to decide better. Customers can either opt-out of reconciliation or change their implementation to match the new requirements. A customer can opt-out of reconciliation by setting the annotation `rbac.authorization.kubernetes.io/autoupdate` to `false` before the upgrade.
```

Comment:
@mfojtik @marun @deads2k want to hear your take on this change and the necessary, explicit opt-out, after we had the trouble with the SCCs breaking customers.

Reply:
> @mfojtik @marun @deads2k want to hear your take on this change and the necessary, explicit opt-out, after we had the trouble with the SCCs breaking customers.

If we follow the recommended approach, you'll have to find a way to count anonymous access to these endpoints *before* trying to tighten it up. 

Also, don't have them prevent reconciliation.  Instead, suggest that they create their own binding.

Reply:
Agree, we need to do to research first, I would like that to be a pre-requirement described in this enhancement and as such include steps how we learn how many clusters in fleet are affected.

Reply:
> Also, don't have them prevent reconciliation. Instead, suggest that they create their own binding.

I understand that upstream already allows that. But I also hesitate to add this to our resources without a good reason.